src/vaadin-text-field-mixin.html

```diff
@@ -482,10 +498,28 @@
 
     /** @private */
     _labelChanged(label) {
-      if (label !== '' && label != null) {
-        this.setAttribute('has-label', '');
-      } else {
-        this.removeAttribute('has-label');
+      this._setOrToggleAttribute('has-label', label === 0 || !!label, this);
+    }
+
+    _helperTextChanged(helperText) {
+      this._setOrToggleAttribute('has-helper', helperText === 0 || !!helperText, this);
+    }
+
+    _errorMessageChanged(errorMessage) {
+      this._setOrToggleAttribute('has-error-message', errorMessage === 0 || !!errorMessage, this);
+    }
+
+    _onHelperSlotChange() {
+      const slottedNodes = this.root.querySelector(`[name="helper"]`).assignedNodes();
```

Comment:
Please use `this.shadowRoot` instead of `this.root` which is the Polymer legacy API.

Reply:
Changed.